tests/msc2716_test.go

```diff
@@ -0,0 +1,222 @@
+// +build msc2716
+
+// This file contains tests for incrementally importing history to an existing room,
+// a currently experimental feature defined by MSC2716, which you can read here:
+// https://github.com/matrix-org/matrix-doc/pull/2716
+
+package tests
+
+import (
+	"fmt"
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/match"
+	"github.com/matrix-org/complement/internal/must"
+	"github.com/tidwall/gjson"
+)
+
+// Test that the message events we insert between A and B come back in the correct order from /messages
+func TestBackfillingHistory(t *testing.T) {
+	deployment := Deploy(t, "rooms_state", b.BlueprintAlice)
+	defer deployment.Destroy(t)
+
+	userID := "@alice:hs1"
+	alice := deployment.Client(t, "hs1", userID)
```

Comment:
Instead of using `alice`, we should be acting as an appservice (`ts` query param usage requires us to be an appservice).

I assume this needs some sort of new blueprint for a homeserver with an appservice registration?

Reply:
Yes, we currently have no support for ASes at all. We would need to:
- Have some way of representing an AS registration in the blueprint.
- Specify what env var / file will be passed into the HS container as AS registration has no HTTP API.

Reply:
@Kegsay What do you think of the current approach to this?

The test doesn't quite pass yet with these application service changes. Getting a 403 when trying to use the `as_token` as the client access token. Need to look into it more.

Reply:
I would like to see AS support as a separate PR. The current approach seems reasonable, but somewhat contorted as ASes don't fit neatly into env vars. It would be better if we just did a volume mount IMO - we already do this for CA functionality. Something along the lines of:
 - `ApplicationService` specified in the blueprint as it currently is
 - These get converted to files from `generateASRegistrationYaml` which then get written to a dedicated volume mount for that particular blueprint/HS tuple.
 - The volume gets mounted when the HS spins up - we mandate that Dockerfiles should inspect this directory e.g `/appservices` on startup and register them before starting.

Reply:
@Kegsay I considered using a volume but there doesn't seem to be a way to add a file to a volume before the Docker container starts. And if the registration file isn't present when Synapse starts up, it throws an error. I can add a `/appservices` volume and copy the same environment variable files there. But it's not clear to me how I can escape the environment variable passing around.

For example trying to do this a little programmatically, I had this before but the Synapse process is already up and running before we can get a handle to interact with the container.

```go
// Create the application service files
for _, as := range hs.ApplicationServices {
	yamlRegistrationContent := generateASRegistrationYaml(as)
	err := d.Docker.CopyToContainer(context.Background(), res.containerID, "/conf/${as.ID}.yaml", strings.NewReader(yamlRegistrationContent), types.CopyToContainerOptions{
		AllowOverwriteDirWithFile: false,
	})

	if err != nil {
		errs = append(errs, err)
		d.log("Failed to add application service registration file %s in %s\n", as.ID, res.containerID)
	}
}
```

One thing I can think of is binding a local directory to the volume and writing to the local directory before starting up the container. But this felt messy to have lying around in the working directory even after we `.gitignore` it.

Please share if I'm missing the obvious solution 😅


---

With the `/ca` cert stuff, we can generate those files on start-up every-time and don't need any config -> https://github.com/matrix-org/complement/blob/f089b60de0bad9af5fcc86edf57056e6ed02035d/dockerfiles/synapse/start.sh#L7-L12

Reply:
> One thing I can think of is binding a local directory to the volume and writing to the local directory before starting up the container.

I was expecting this tbh - this is what the CA stuff does iirc - see https://github.com/matrix-org/complement/blob/master/internal/docker/builder.go#L394

Reply:
Thanks @Kegsay! Got things working and split the application service support out into https://github.com/matrix-org/complement/pull/74

```
--- PASS: TestBackfillingHistory (6.91s)
    --- PASS: TestBackfillingHistory/parallel (0.00s)
        --- PASS: TestBackfillingHistory/parallel/Backfilled_messages_come_back_in_correct_order (0.06s)
        --- PASS: TestBackfillingHistory/parallel/Backfilled_events_without_m.historical_come_down_/sync (0.95s)
        --- PASS: TestBackfillingHistory/parallel/Backfilled_events_with_m.historical_do_not_come_down_/sync (0.97s)
PASS
```